### ISSUE TITLE
New test: the avalanche criterion

### ISSUE THREAD

**Issue by avaneev** [rurban/smhasher#145]
This test is best used in Words test. It calculates the average bit flip count in pair-wise hash comparisons, demonstrating the closeness to the strict avalanche criterion (50.0%).

Example code:
```
	const int HashLen = 32;
	uint8_t Hash1[ HashLen ];
	uint8_t Hash2[ HashLen ];
	uint64_t bc = 0;

	for( i = 0; i < StringList.getItemCount() - 1; i++ )
	{
		prvhash42s_oneshot( (uint8_t*) VOXSTR( StringList[ i ]),
			StringList[ i ].getLength(), Hash1, HashLen );

		prvhash42s_oneshot( (uint8_t*) VOXSTR( StringList[ i + 1 ]),
			StringList[ i + 1 ].getLength(), Hash2, HashLen );

		bc += countDiffBits( Hash1, Hash2, HashLen );
	}

	printf( "Avalanche criterion: %.1f%%\n", 100.0 * bc /
		( StringList.getItemCount() - 1 ) / ( HashLen * 8 ));
```

```
inline int countDiffBits( const uint8_t* const Hash1,
	const uint8_t* const Hash2, const int HashLen )
{
	int cc = 0;
	int m;

	for( m = 0; m < HashLen; m++ )
	{
		uint8_t v1 = Hash1[ m ];
		uint8_t v2 = Hash2[ m ];
		int j;

		for( j = 0; j < 8; j++ )
		{
			cc += ( v1 & 1 ) ^ ( v2 & 1 );
			v1 >>= 1;
			v2 >>= 1;
		}
	}

	return( cc );
}
```

**Comment by avaneev** [comment]
@rurban A note on usage of hashes for file digests: usually the disk and network IO will be much slower than a hash function, so practically a 6GB/s (48Gbit/s) hash function is plenty enough. I would focus on cryptographic proofs of hash functions: the Avalanche criterion is a part of the proof, as well as the "bit frequency normalcy" test I've proposed before. The `sparse long` test is good at catching broken hash functions: I did use it myself to catch bugs. For example, a hash function may pass the Words test, but it may fail the `sparse long` test easily. Most mul+xorshift hash functions are irreversible anyway.
